performance_manager/lib/s3_utils.py

```diff
@@ -5,30 +5,72 @@
 import pandas
 import pyarrow.parquet as pq
 from pyarrow import fs
+import pyarrow
 
 
-def read_parquet(
+def _get_pyarrow_table(
     filename: Union[str, List[str]],
-    columns: Union[List[str], slice] = slice(None),
     filters: Optional[Union[Sequence[Tuple], Sequence[List[Tuple]]]] = None,
-) -> pandas.core.frame.DataFrame:
+) -> pyarrow.Table:
     """
-    read parquet file or files from s3 and return it as a pandas dataframe
+    internal function to get pyarrow table from parquet file(s)
     """
     active_fs = fs.S3FileSystem()
     if isinstance(filename, list):
         to_load = [f.replace("s3://", "") for f in filename]
     else:
         to_load = [filename.replace("s3://", "")]
 
+    # using `read_pandas` because `read` with "columns" parameter results in
+    # much slower file downloads for some reason...
+    return pq.ParquetDataset(
+        to_load, filesystem=active_fs, filters=filters
+    ).read_pandas()
+
+
+def read_parquet(
+    filename: Union[str, List[str]],
+    columns: Union[List[str], slice] = slice(None),
+    filters: Optional[Union[Sequence[Tuple], Sequence[List[Tuple]]]] = None,
+) -> pandas.core.frame.DataFrame:
+    """
+    read parquet file or files from s3 and return it as a pandas dataframe
+    """
     return (
-        pq.ParquetDataset(to_load, filesystem=active_fs, filters=filters)
-        .read_pandas()
-        .to_pandas()
+        _get_pyarrow_table(filename, filters)
+        .to_pandas(self_destruct=True)
         .loc[:, columns]
     )
 
 
+def read_parquet_chunks(
+    filename: Union[str, List[str]],
+    max_rows: int = 100_000,
+    columns: Union[List[str], slice] = slice(None),
+    filters: Optional[Union[Sequence[Tuple], Sequence[List[Tuple]]]] = None,
+) -> Iterator[pandas.core.frame.DataFrame]:
+    """
+    read parquet file or files from s3 IN CHUNKS
+    return chunks as pandas dataframes
+
+    chunk size attempts to be close to max_rows parameter, but may sometimes
+    overshoot because of chunk layout of pyarrow table
+    """
+    yield_dataframe = pandas.DataFrame()
+    for batch in _get_pyarrow_table(filename, filters).to_batches(
+        max_chunksize=None
+    ):
```

Comment:
interesting to me that 
```
    return pq.ParquetDataset(
        to_load, filesystem=active_fs, filters=filters
    ).read_pandas()
```
doesn't load the whole thing.

Reply:
I'm not totally sure if it actually loads the pyarrow table at that point or not, it might lazy load the table/schema and wait for the `'to_pandas()` method to actually download.

Even if it does download the pyarrow table on that call, the memory usage is a fraction of converting the table to a dataframe.